lib/PrivateAppUserTokenManager.ts

```diff
@@ -0,0 +1,243 @@
+        accountId: this.accountId,
+        appId: appId,
+      });
+    }
+    this.tokenMap.set(appId, token);
+    if (this.tokenMapIntervalId.has(appId)) {
+      clearInterval(this.tokenMapIntervalId.get(appId));
+    }
+    const now = moment.utc();
+    const refreshDelayMillis = Math.max(
+      moment
+        .utc(token.expiresAt)
+        .subtract(5, 'minutes')
+        .diff(now, 'milliseconds'),
+      0
+    );
+    this.tokenMapIntervalId.set(
+      appId,
+      setInterval(
+        () => this.refreshToken(appId, token.userTokenKey, scopeGroups),
+        refreshDelayMillis
+      )
+    );
+    logger.debug(
+      i18n(`${i18nKey}.refreshScheduled`, {
+        appId: appId,
+        expiresAt: token.expiresAt,
+        refreshTime: now.add(refreshDelayMillis, 'milliseconds').toISOString(),
+      })
+    );
+  }
+
+  private async createOrGetActiveToken(
+    appId: number,
+    scopeGroups: string[]
+  ): Promise<PrivateAppUserTokenResponse> {
+    const maybeToken = await this.getExistingToken(appId);
+    if (maybeToken === null) {
+      logger.debug(
+        i18n(`${i18nKey}.create`, {
+          appId: appId,
+        })
+      );
+      return createPrivateAppUserToken(this.accountId, appId, scopeGroups);
+    } else if (
+      moment
+        .utc()
+        .add(5, 'minutes')
+        .isAfter(moment.utc(maybeToken.expiresAt)) ||
+      !PrivateAppUserTokenManager.doesTokenHaveAllScopes(
+        maybeToken,
+        scopeGroups
+      )
+    ) {
+      logger.debug(
+        i18n(`${i18nKey}.refresh`, {
+          appId: appId,
+        })
+      );
+      return updatePrivateAppUserToken(
+        this.accountId,
+        appId,
+        maybeToken.userTokenKey,
+        scopeGroups
+      );
+    }
+    return maybeToken;
+  }
+
+  private static doesTokenHaveAllScopes(
+    token: PrivateAppUserTokenResponse | undefined,
+    requestedScopes: string[]
+  ): boolean {
+    if (token === undefined) {
+      return false;
+    }
+    return requestedScopes.every(scopeGroup =>
+      token.scopeGroups.includes(scopeGroup)
+    );
+  }
+
+  private async getExistingToken(
+    appId: number
+  ): Promise<PrivateAppUserTokenResponse | null> {
+    try {
+      return await fetchPrivateAppUserToken(this.accountId, appId);
+    } catch (err) {
+      if (isAxiosError(err)) {
+        if (err.response?.status == 404) {
+          return null;
+        }
+      }
+      throw err;
+    }
+  }
+
+  private async refreshToken(
+    appId: number,
+    userTokenKey: string,
+    scopeGroups: string[]
+  ) {
+    logger.debug(
+      i18n(`${i18nKey}.refresh`, {
+        appId: appId,
+      })
+    );
+    const newToken = await updatePrivateAppUserToken(
+      this.accountId,
+      appId,
+      userTokenKey,
+      scopeGroups
+    );
+    if (newToken) {
+      this.setCacheAndRefresh(appId, newToken, scopeGroups);
+    } else {
+      throwErrorWithMessage(`${i18nKey}.errors.refreshFailed`, {
```

Comment:
How does this thrown error impact the running process? Will it kill the local dev command? This would prevent future refreshes. I wonder if we should consider adding in a retry here since a failure would be pretty bad. I think we'd at least want to inform the user that the temporary PAT isn't working anymore.

Reply:
See my comment above on the en.lyaml file. Should we give the user some instructions to troubleshoot?